src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,338 @@
+    if (parameters.getCancellationType() == ActivityCancellationType.ABANDON) {
+      notifyCanceled();
+    } else {
+      action(Action.CANCEL);
+    }
+  }
+
+  private void cancelScheduleCommand() {
+    cancelInitialCommand();
+    if (parameters.getCancellationType() != ActivityCancellationType.ABANDON) {
+      notifyCanceled();
+    }
+  }
+
+  private void notifyCanceledIfTryCancel() {
+    if (parameters.getCancellationType() == ActivityCancellationType.TRY_CANCEL) {
+      notifyCanceled();
+    }
+  }
+
+  private void notifyCanceled() {
+    Failure canceledFailure =
+        Failure.newBuilder()
+            .setSource(JAVA_SDK)
+            .setCanceledFailureInfo(CanceledFailureInfo.getDefaultInstance())
+            .build();
+    completionCallback.apply(Optional.empty(), canceledFailure);
+  }
+
+  private void notifyCompleted() {
+    ActivityTaskCompletedEventAttributes completedAttr =
+        currentEvent.getActivityTaskCompletedEventAttributes();
+    Optional<Payloads> result =
+        completedAttr.hasResult() ? Optional.of(completedAttr.getResult()) : Optional.empty();
+    completionCallback.apply(result, null);
+  }
+
+  private void notifyFailed() {
+    ActivityTaskFailedEventAttributes failed = currentEvent.getActivityTaskFailedEventAttributes();
+    ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+    ActivityFailureInfo failureInfo =
+        ActivityFailureInfo.newBuilder()
+            .setActivityId(scheduleAttr.getActivityId())
+            .setActivityType(scheduleAttr.getActivityType())
+            .setIdentity(failed.getIdentity())
+            .setRetryState(failed.getRetryState())
+            .setScheduledEventId(failed.getScheduledEventId())
+            .setStartedEventId(failed.getStartedEventId())
+            .build();
+    Failure failure =
+        Failure.newBuilder()
+            .setActivityFailureInfo(failureInfo)
+            .setCause(failed.getFailure())
+            .setMessage("Activity task failed")
+            .build();
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyTimedOut() {
+    ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+
+    ActivityTaskTimedOutEventAttributes timedOut =
+        currentEvent.getActivityTaskTimedOutEventAttributes();
+
+    ActivityFailureInfo failureInfo =
+        ActivityFailureInfo.newBuilder()
+            .setActivityId(scheduleAttr.getActivityId())
+            .setActivityType(scheduleAttr.getActivityType())
+            .setRetryState(timedOut.getRetryState())
+            .setScheduledEventId(timedOut.getScheduledEventId())
+            .setStartedEventId(timedOut.getStartedEventId())
+            .build();
+    Failure failure =
+        Failure.newBuilder()
+            .setActivityFailureInfo(failureInfo)
+            .setCause(timedOut.getFailure())
+            .setMessage("Activity task timedOut")
```

Comment:
should we have string constants for these so that its easy for a technical writer to review or so we can do translations (in the future)

Reply:
Would you file an issue for this? I think it is a great idea, but we want to do it holistically with the whole SDK.

Reply:
Hopefully we will never do translations :-)